src/main/kotlin/org/rust/ide/annotator/RsFormatMacroAnnotator.kt

```diff
@@ -13,124 +13,123 @@ import com.intellij.openapiext.isUnitTestMode
+        // skip advanced checks and highlighting if there are syntax errors
+        if (errors.isNotEmpty()) {
+            return
+        }
+
+        highlightParametersInside(parseCtx, holder)
+
+        if (!isUnitTestMode && element.project.macroExpansionManager.macroExpansionMode !is MacroExpansionMode.New) return
+
+        val parameters = buildParameters(parseCtx)
+        val arguments = macroArgs
+            .drop(macroPos + 1)
+            .toList()
+        val ctx = FormatContext(parameters, arguments, formatMacro)
+
+        val annotations = checkParameters(ctx).toMutableList()
+        annotations += checkArguments(ctx)
+
+        for (annotation in annotations) {
+            holder.createErrorAnnotation(annotation.range, annotation.error)
         }
     }
+}
 
-    private fun macroToFormatPos(macro: String): Int = when (macro) {
-        "println" -> 0
-        "print" -> 0
-        "eprintln" -> 0
-        "eprint" -> 0
-        "format" -> 0
-        "format_args" -> 0
-        "format_args_nl" -> 0
-        "write" -> 1
-        "writeln" -> 1
-        else -> -1
+private data class ParameterMatchInfo(val range: TextRange, val match: MatchResult, val type: String)
+
+private sealed class FormatParameter(val matchInfo: ParameterMatchInfo) {
+    override fun toString(): String {
+        return this.matchInfo.match.value
+    }
+
+    val range: TextRange = matchInfo.range
+
+    fun requiredTrait(knownItems: KnownItems): RsTraitItem? {
+        return when (this.matchInfo.type) {
+            "" -> knownItems.Display
+            "?", "x?", "X?" -> knownItems.Debug
+            "o" -> knownItems.Octal
+            "x" -> knownItems.LowerHex
+            "X" -> knownItems.UpperHex
+            "p" -> knownItems.Pointer
+            "b" -> knownItems.Binary
+            "e" -> knownItems.LowerExp
+            "E" -> knownItems.UpperExp
+            else -> return null
+        }
     }
+}
+private class NamedParameter(matchInfo: ParameterMatchInfo, val name: String) : FormatParameter(matchInfo)
+private class PositionalParameter(matchInfo: ParameterMatchInfo, val position: Int) : FormatParameter(matchInfo)
+
+private data class FormatContext(val parameters: List<FormatParameter>,
+                                 val arguments: List<RsFormatMacroArg>,
+                                 val macro: RsMacroCall) {
+    val namedParameters: Set<NamedParameter> = parameters.filterIsInstance<NamedParameter>().toSet()
+    val positionalParameters: List<PositionalParameter> = parameters.filterIsInstance<PositionalParameter>().toList()
+
+    val namedArguments: Set<RsFormatMacroArg> = arguments.filter { it.name() != null }.toSet()
+
+    val knownItems: KnownItems = macro.knownItems
+}
+
+private data class ParsedParameter(val completeMatch: MatchResult,
+                                   val innerContentMatch: MatchResult? = null) {
+    val innerContent = completeMatch.groups[2]
+    val range: IntRange = completeMatch.range
```

Comment:
Looks like `range` is mapped in all usages (`ctx.toSourceRange(parameter.range)`), so maybe it's possible to map it `toSourceRange` once?

Reply:
It turned out that once I added the new width/precision specifier parsing, this method became more useful, so I kept it like this.